p2p/net/swarm/dial_worker.go

```diff
@@ -149,50 +227,69 @@ loop:
 				continue loop
 			}
 
-			// the request has some pending or new dials, track it and schedule new dials
+			// The request has some pending or new dials. We assign this request a request number.
+			// This value of w.reqno is used to track this request in all the structures
 			w.reqno++
-			w.requests[w.reqno] = pr
+			w.pendingRequests[w.reqno] = pr
 
 			for _, ad := range tojoin {
-				if simConnect, isClient, reason := network.GetSimultaneousConnect(req.ctx); simConnect {
-					if simConnect, _, _ := network.GetSimultaneousConnect(ad.ctx); !simConnect {
-						ad.ctx = network.WithSimultaneousConnect(ad.ctx, isClient, reason)
+				if !ad.dialed {
+					// we haven't dialed this address. update the ad.ctx to have simultaneous connect values
+					// set correctly
+					if simConnect, isClient, reason := network.GetSimultaneousConnect(req.ctx); simConnect {
+						if simConnect, _, _ := network.GetSimultaneousConnect(ad.ctx); !simConnect {
+							ad.ctx = network.WithSimultaneousConnect(ad.ctx, isClient, reason)
+							// update the element in dq to use the simultaneous connect delay.
+							dq.Add(network.AddrDelay{Addr: ad.addr, Delay: addrDelay[ad.addr]})
+						}
 					}
 				}
+				// add the request to the addrDial
 				ad.requests = append(ad.requests, w.reqno)
 			}
 
 			if len(todial) > 0 {
+				// these are new addresses, track them and add them to dq
 				for _, a := range todial {
-					w.pending[a] = &addrDial{addr: a, ctx: req.ctx, requests: []int{w.reqno}}
+					w.trackedDials[a] = &addrDial{addr: a, ctx: req.ctx, requests: []int{w.reqno}}
+					dq.Add(network.AddrDelay{Addr: a, Delay: addrDelay[a]})
 				}
-
-				w.nextDial = append(w.nextDial, todial...)
-				w.nextDial = w.rankAddrs(w.nextDial)
-
-				// trigger a new dial now to account for the new addrs we added
-				w.triggerDial = ready
 			}
-
-		case <-w.triggerDial:
-			for _, addr := range w.nextDial {
+			// setup dialTimer for updates to dq
+			scheduleNextDial()
+
+		case <-dialTimer.Ch():
+			// It is time to dial the most urgent addresses.
+			// We dont check the delay here because an early trigger means all in flight
+			// dials have completed.
```

Comment:
What does that mean? What's an early trigger, how would that happen?

Reply:
Say we are dialing a quic address(q1) and the second quic address(q2) is delayed by 250ms. 
Now if q1 fails after 100ms, we will dial q2 immediately rather than waiting for 150ms more. 

When q1 fails, we receive the error notification on w.resch and decrement dialsInFlight to 0. 
then [scheduleNextDial](https://github.com/sukunrt/go-libp2p/blob/smart-dialing/p2p/net/swarm/dial_worker.go#L135) will reset the timer to trigger immediately, and so we'll dial q2 immediately.